lib/acme_server/jobs/http_verifier.ex

```diff
@@ -1,23 +1,49 @@
 defmodule AcmeServer.Jobs.HttpVerifier do
+  @moduledoc false
+
+  # This module powers a single http verifier job, which issues an http challenge
+  # to the server. If the challenge succeeds, the job updates the account info.
+  #
+  # Each verifier is running as a separate process, which ensure proper error
+  # isolation. Failure or blockage of while verifying one site won't affect
+  # other verifications.
+  #
+  # A verifier process is a Parent.GenServer which starts the actual verification
+  # as a child task. This approach is chosen for better control with error
+  # handling. The parent process can apply delay and retry logic, and give
+  # up after some number of retries.
+  #
+  # Because failure of one verification shouldn't affect others, the restart
+  # strategy is temporary. In principle, the Parent.GenServer has minimal logic,
+  # since most of the action is happening in the child task, so it shouldn't
+  # crash. But even if it does, we don't want to trip up the restart intensity,
+  # and crash other verifiers.
+
+  # We'll retry at most 12 times, with 5 seconds delay between each retry.
+  # Each verification request must return in 5 seconds. Therefore, in total
+  # we're verifying for at most 1 minute 55 seconds (12 timeouts of 5 seconds
+  # plus 11 delays of 5 seconds).
+  @max_retries 12
```

Comment:
Is this more robust than using the child spec to explicitly set the number of retries?  I am not an erlang expert, so pragmatic fault tolerance isn’t a forte of mine.

Reply:
With supervisors you can't even implement this. In the supervisor you can provide max number of restarts and the time interval. However, that setting is for all children, not a single child. Therefore, if you e.g. use `max_restarts: 2, max_seconds: 1`, two independent failures will cause the supervisor to be restarted. That in turn will take down all pending challenges, not just the offending one.

You could work around this by making each challenge task be supervised by it's own supervisor. So you have one common supervisor, and each child is another supervisor (where you set restart options), and then a single child of that supervisor is a task. However, that won't give you the retry delay, so you'll need to do some extra improvisations to make that happen.

This doesn't mean that supervisors are bad. They are great at what they do, which is organizing hierarchy, allowing proper startup and shutdown order, isolating failures, and helping the system self-heal from heisenbugs by restarting the minimal part of the system. 

However, the network failure scenario is not a bug, it's an expected situation which needs to be handled in a particular way (sleep awhile, try again), and supervisors are just not expressive enough for that.